Documentation/design-docs/COM-activation.md

```diff
@@ -0,0 +1,199 @@
+
+During .NET class registration, the shim is identified as the in-proc server for the class. Additional metadata is inserted into the registry to indicate what .NET assembly to load and what type to activate. For example, in addition to the typical [in-proc server](https://docs.microsoft.com/en-us/windows/desktop/com/inprocserver32) registry values the following values are added to the registry for the `TypeLoadException` class.
+
+```
+"Assembly"="mscorlib, Version=1.0.5000.0, Culture=neutral, PublicKeyToken=b77a5c561934e089"
+"Class"="System.TypeLoadException"
+"RuntimeVersion"="v1.1.4322"
+```
+
+The above registration is typically done with the [`RegAsm.exe`](https://docs.microsoft.com/en-us/dotnet/framework/tools/regasm-exe-assembly-registration-tool) tool. Alternatively, registry scripts can be generated by `RegAsm.exe`.
+
+### .NET Core Class COM Activation
+
+In .NET Core, our intent will be to avoid a system wide shim library. This decision may add additional cost for deployment scenarios, but will reduce servicing and engineering costs by making deployment more explicit and less magic.
+
+The current .NET Core hosting solutions are described in detail at [Documentation/design-docs/host-components.md](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/host-components.md). Along with the existing hosts an additional activation host library will be added. This library (henceforth identified as 'shim') will export the required functions for COM class activation and registration and act in a way similar to `mscoree.dll`.
+
+>[`HRESULT DllGetClassObject(REFCLSID rclsid, REFIID riid, LPVOID *ppv);`](https://docs.microsoft.com/en-us/windows/desktop/api/combaseapi/nf-combaseapi-dllgetclassobject)
+
+>[`HRESULT DllCanUnloadNow();`](https://docs.microsoft.com/en-us/windows/desktop/api/combaseapi/nf-combaseapi-dllcanunloadnow)
+
+>[`HRESULT DllRegisterServer();`](https://msdn.microsoft.com/en-us/library/windows/desktop/ms682162(v=vs.85).aspx)
+
+>[`HRESULT DllUnregisterServer();`](https://msdn.microsoft.com/en-us/library/windows/desktop/ms691457(v=vs.85).aspx)
+
+When `DllGetClassObject()` is called in a COM activation scenario, the following will occur:
+
+1) Determine additional registration information needed for activation.
+    * The shim will check for an embedded manifest. If the shim does not contain an embedded manifest, the shim will check if a file with the `<shim_name>.clsidmap` naming format exists adjacent to it. Build tooling will be expected to handle shim customization, including renaming the shim to be based on the managed assembly's name (e.g. `NetComServer.dll` could have a custom shim called `NetComServer.shim.dll`).
```

Comment:
Why also allow a sidecar? It wouldn't be possible to ensure a sidecar was signed.

Reply:
> the dev cycle is much more complicated

Which part of the dev cycle is much more complicated? The manifest handling will be hidden behind msbuild target. I assume that this msbuild target will embed it by default.

Reply:
The following will be more complicated:

* Incremental build - MSBuild does handle a lot of this but there are many edge cases around auto-generated files that may crop up depending on how we do project references or where in the pipeline we create the manifest.

* Single shim multiple managed assemblies - When multiple managed binaries that proffer COM classes we will need to coordinate with the shim customizing target to ensure all of them can be hosted by the same shim. We are going to want the option for a single shim for multiple managed assemblies for size sensitive systems.

* Diagnostics for the pipeline - Instead of hiding the entire system if we make the pipeline a series of well-known and defined steps/targets we can optimized it by making a single MSBuild target, but we can also help users to help themselves and point at the loose file option when there is an issue. This also enables a reasonable testing of the pipeline.

I do not want to make magic in COM activation for .NET Core. Requiring embedding or hiding the manifest turns into magic - which is fine when it works, but we will have bugs and allowing an escape hatch for customer should be our primary goal - unblocking user scenarios.

> I assume that this msbuild target will embed it by default.

I agree that is a reasonable default.

Reply:
If we support embedding, we need to build and test all what you have listed and it needs to better works well if it is the default. Having second option is just more work, more tests and more documentation on top of it. I think we should be introducing the second options here only if we get feedback that the one embedding option does not work well.

We can drill into details on the incremental build, multiple assemblies and diagnostic concerns once we get to the tooling design and implementation.